lib/elixir/unicode/security.ex

```diff
@@ -106,7 +106,133 @@ defmodule String.Tokenizer.Security do
     #    the specified data, producing a string of exemplar characters.
     #  - Reapply NFD." (UTS 39 section 4, skeleton definition)
     :unicode.characters_to_nfd_list(s)
-    |> Enum.map(&confusable_prototype/1)
+    |> bidi_skeleton()
     |> :unicode.characters_to_nfd_list()
   end
+
+  # unicode 15 adds bidiSkeleton because, w/RTL codepoints, idents that
+  # aren't confusable LTR *are* confusable in most places human review
+  # occurs (editors/browsers, thanks to bidi algo, UAX9).
+  #
+  # The solution is to detect spans with reversed visual direction,
+  # and reverse those, so that the input we check for confusability
+  # matches the perceived sequence instead of the byte sequence.
+  #
+  # (we need this regardless of script mixing, because direction-neutral
+  # chars like _ or 0..9 can mix w/RTL chars).
+  def bidi_skeleton(s) do
+    # UTS39-28 4:
+    #  'Bidirectional confusability is costlier to check than
+    #   confusability, as [unicode bidi algo] must be applied.
+    #   [...] a fast path can be used: [...] if X has no characters
+    #   w/bidi classes R or AL, bidiSkeleton(X) = skeleton(X)
+    #
+    if length(s) > 1 and any_rtl?(s) do
+      unbidify(s) |> Enum.map(&confusable_prototype/1)
+    else
+      Enum.map(s, &confusable_prototype/1)
+    end
+  end
+
+  # load direction-changing and neutral codepoints valid in idents
+  {rtls, neutrals} =
+    Path.join(__DIR__, "UnicodeData.txt")
+    |> File.read!()
+    |> String.split(["\r\n", "\n"], trim: true)
+    |> Enum.reduce({[], []}, fn line, {rtls, neutrals} = acc ->
+      with [point, _, _, _, bidi | _] <- String.split(line, ";"),
+           codepoint = String.to_integer(point, 16),
+           true <- String.Tokenizer.id_codepoint?(codepoint) do
+        # https://www.unicode.org/reports/tr44/tr44-32.html#Bidi_Class_Values
+        cond do
+          bidi in ~w(R AL)s -> {[codepoint | rtls], neutrals}
+          bidi in ~w(WS ON CS EN ES ET NSM)s -> {rtls, [codepoint | neutrals]}
+          true -> acc
+        end
+      else
+        _ -> acc
+      end
+    end)
+
+  rangify = fn [head | tail] ->
+    {first, last, acc} =
+      Enum.reduce(tail, {head, head, []}, fn
+        number, {first, last, acc} when number == first - 1 ->
+          {number, last, acc}
+
+        number, {first, last, acc} ->
+          {number, number, [{first, last} | acc]}
+      end)
+
+    [{first, last} | acc]
+  end
+
+  # direction of a codepoint. (rtl, neutral, ltr fallback)
+  for {first, last} <- rangify.(rtls) do
```

Comment:
Let's define this function in String.Tokenizer itself, so we don't have to parse UnicodeData additional times.